sig-testing/flakiness-sla.md

```diff
@@ -0,0 +1,84 @@
+
+### Measuring flakiness
+There are a number of challenges in monitoring flakiness. We expect that
+the metric will be heuristic in nature and we will iterate on it over time.
+Identifying all of the potential problems and ways to measure the metric
+are out of the scope of the document, but some currently known challenges
+are listed at the end of the document.
+
+## Flakiness SLA
+We will measure flakiness based on pull requests that are run through pre-submit
+PR testing. The metric that we will monitor is: Pre-submit flakes per PR. This metric
+will be calculated on a daily and weekly basis using:
+
+Sum(Flakes not caused by the PR) / Total(PRs)
+
+Our current SLA is that this metric will be less than 0.01
+(1% of all PRs have flakes that are not caused by the PR itself).
+
+## Activities on SLA violation
+When the project is out of SLA for flakiness on either the daily or weekly metric
+we will stop the merge queue.
+
+Only PRs with the label "flake-fix" will be allowed to merge. The merge queue
+will remain blocked until a sufficient number of flake-fixes have been
+submitted to restore the daily CLA.
+
+### Exceptions
+We may consider merging PRs under exceptional circumstances, for example
+merging security fixes or other critical bug fixes that need to be cut into
+a pre-existing release. Each exception will be considered on a case by case
+basis. Exceptions will only be granted for PRs that are already labelled with
+cherry-pick approve.
+
+## Non-goals: Flakiness at HEAD
+We will _not_ currently measure flakiness for continuous e2e that
+are run at HEAD.
+
+There are a number of reasons for this:
+   * We already block the submit-queue if certain HEAD e2e tests are failing
+   * The rate at which code is merged at HEAD is so great that its rare that a test will be re-run for the exact same git commit.
+   * The volume of testing on PRs is significantly higher than at HEAD, so flakes are more readily apparent.
+   * The flakiness of these tests is already being measured in other places.
+
+## Challenges in measuring the flakiness metric
+
+Currently, measuring this metric is somewhat tricky, since determining flakes
+caused by PRs vs. existing flakes is somewhat challenging. We will use a variety
+of heuristics to determine this, including looking at PRs that contain changes
+that could not cause flakes (e.g. typos, docs). As well as looking at the
+past history of test failures (e.g. if a test fails across many different PRs
+it's likely a flake)
+
+### Changes in the environment (e.g. GCE, Azure, AWS, etc)
+Changes in the environment are notably hard to measure or control for, but we'll do our best.
+
+### Changes in code
+Changes in code are obviously quite easy to control.
+
+### Caused by PR or pre-existing?
+Measuring whether a flake is caused by a PR or existing problems is a challenge, but will use
+observation across multiple PRs to judge the probability that it is an existing problem or caused by the PR.
```

Comment:
@bgrant0607 
Thanks I mentioned in in comments, but I didn't add it to the docs, added.

@derekwaynecarr 
I don't think we should only test lgtm + approve, I'm not sure what that signal tells us in this context.

@liggitt 
what are your concerns about blocking the queue on unmerged code?

Right now, 95% (or more) of the flakes that we see on PRs are due to problems in the tests not in the unmerged code.

Additionally, we will look for tests that are failing across multiple different PRs, which further reduces the chances that a flake is caused by the particular PR.

It seems pretty clear to me that we can detect flakes quite reliably. As I mentioned before, if we'd prefer to start with an expectation of a 5% flake rate as we iterate on the metric, I'm ok with that.